## 1.1. Sequência de Fibonacci

A Sequência de Fibonacci é uma sequência numérica de tamanho **N** onde qualquer número é dado pela soma dos dois elementos anteriores, com exeção do primeiro e do segundo número.


### 1.1.1. Primeira tentativa com recursão

In [ ]:
def fibonacci_1(n: int) -> int:
    return fibonacci_1(n - 1) + fibonacci_1(n - 2)

fibonacci_1(5)

: 

: 

O programa quebra pois não há uma condição de parada, então acabamos estourando a pilha de recursão do Python.

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in fib
  File "<stdin>", line 2, in fib
  File "<stdin>", line 2, in fib
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded
```

Para solucionar isso, devemos trabalhar com uma condição de parada para a recursão.

### 1.1.2. Utilizando casos de base

O caso base de uma função recursiva é uma condição onde a função não chamará a si mesma novamente, quebrando assim a recursão e impedindo que aconteça um _loop infinito_.

In [16]:
fib_2_execution_count: int = 0

def fibonacci_2(n: int) -> int:
    global fib_2_execution_count
    fib_2_execution_count += 1
    # caso base que para a recursão
    if n < 2:
        return n
    # chamada recursiva
    return fibonacci_2(n - 2) + fibonacci_2(n - 1)

n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_2(n)}')
print(f'Para calculá-lo foram necessárias {fib_2_execution_count} chamadas a função fibonacci_2')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 2692537 chamadas a função fibonacci_2


### 1.1.3. Memoização

A memoização consiste em armazenar os resultados gerados por uma função, de forma que toda vez que formos executá-la novamente, caso os dados de entrada já tenham sido processados anteriormente, será retornado o valor armazenado, caso contrário a função será executada normalmente e ao final salvará a saída gerada para aquela entrada.

In [17]:
from typing import Dict

fib_3_execution_count: int = 0

# iniciamos nosso dicionário com os dois casos base da sequência de fibonacci
memo: Dict[int, int] = {0: 0, 1: 1}

def fibonacci_3(n: int) -> int:
    global fib_3_execution_count
    fib_3_execution_count += 1
    if n not in memo:
        memo[n] = fibonacci_3(n - 2) + fibonacci_3(n - 1)
    return memo[n]

n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_3(n)}')
print(f'Para calculá-lo foram necessárias {fib_3_execution_count} chamadas a função fibonacci_3')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 59 chamadas a função fibonacci_3


Ao utilizar a memoização, durante o cálculo do trigésimo número da sequência saimos de `2692537` chamadas a função `fibonacci_2` para apenas `59` chamadas a função `fibonacci_3`.